src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -408,9 +408,47 @@ private class Emit(
           xvout := xmout.mux(defaultValue(typ), xvout)
         ), xmout, xvout)
 
-      case x@ApplyAggOp(a, op, args) =>
-        val agg = AggOp.get(op, x.inputType, args.map(_.typ))
-        present(emitAgg(a)(agg.seqOp(aggregator, _, _)))
+      case ArrayFor(a, valueName, body) =>
+        val tarray = coerce[TArray](a.typ)
+        val eti = typeToTypeInfo(tarray.elementType)
+        val xmv = mb.newField[Boolean]()
+        val xvv = coerce[Any](mb.newField(valueName)(eti))
+        val bodyenv = env.bind(
+          valueName -> (eti, xmv.load(), xvv.load()))
+        val codeB = emit(body, env = bodyenv)
+        val aBase = emitArrayIterator(a)
+        val cont = { (m: Code[Boolean], v: Code[_]) =>
+          Code(
+            xmv := m,
+            xvv := v,
+            codeB.setup)
+        }
+
+        val processAElts = aBase.arrayEmitter(cont)
+        val ma = processAElts.m.getOrElse(const(false))
+        EmitTriplet(
+          Code(
+            processAElts.setup,
+            ma.mux(
+              Code._empty,
+              Code(aBase.calcLength, processAElts.addElements))),
+          const(false),
+          Code._empty)
+
+      case SeqOp(a, i, agg) =>
+        EmitTriplet(
+          emitAgg(a, env)(agg.seqOp(aggregator(i), _, _)),
+          const(false),
+          Code._empty)
+
+      case Begin(xs) =>
+        EmitTriplet(
+          coerce[Unit](Code(xs.map { x =>
+            val codex = emit(x)
+            codex.setup
+          }: _*)),
+          const(false),
+          Code._empty)
```

Comment:
It seems wrong to not also emit the `v`.

We're in a brave new world, and I haven't kept up with IR changes, but I was thinking the semantics was:

 - setup must be executed exactly once before `v` and `m` are executed
 - at least `v` must be laid down if you want to "evaluate" the ir

Reply:
It's weird now that we have side-effects...

Reply:
> at least v must be laid down if you want to "evaluate" the ir

I think Code[Unit] side effects should be in v.  You shouldn't have to execute v if you don't want to.  For example, `IsNA` doesn't evaluate v because it doesn't care about the value.

I guess I think setup should have all the effects.